woocommerce/woocommerce-ios
# Add Product: present product type bottom sheet then navigate to product form

Create a coordinator for Add Product navigation #2740 

## Changes

- Created `ProductFactory` that generates a new `Product` given a site ID and product type
- Created `AddProductCoordinator` that implements `Coordinator` (I decided to go with `Coordinator` instead of `CoordinatorController: UINavigationController` because this navigation flow does not use a new navigation controller)
  - First, it presents a product type bottom sheet (note that there's a current issue with `UIBarButtonItem` anchor on iPad that I'll address in a separate subtask that involves updating a shared framework `WordPressUI`)
  - Once a product type is selected, it creates an empty product via `ProductFactory` and navigates to a product form `ProductFormViewController`
- In `ProductsViewController`, implemented `addProduct` that starts `AddProductCoordinator` navigation
- Added one test case for `AddProductCoordinator`, but very limited to the presented bottom sheet since the content of the bottom sheet is private. Please lemme know if you can think of other ways to test the navigation further!

## Testing

Note that there are still UI/UX differences that are to be implemented in a separate subtask in #2740 (ref p91TBi-3f7-p2), so that a product can only be created after making a change from the default empty product. Please also test on an iPhone simulator/device since there's another subtask to fix the bottom sheet anchoring on a `UIBarButtonItem`.

I noticed a slight delay after selecting a product type from the bottom sheet and before the product form navigation on iPhone 11 iOS 13.7 simulator. I tested on an iOS 13.7 device and iOS 12.1 simulator and didn't see the delay. Time profiler also didn't show anything suspicious during the delay, so it might be an iOS 13 simulator issue.

- Go to the Products tab
- Tap on the "+" CTA --> a bottom sheet of Simple, Grouped, and External product types should be shown (no Variable)
- Select a product type --> the bottom sheet should be dismissed and then it navigates to the product details
- Make some changes to the new product (e.g. adding an image, updating title, adding price/shipping/inventory info)
- Tap "Update" (this will be updated to "Publish" / "Save as draft" in the future) --> the product should be added remotely

## Example screenshots

Please note that there are still UI/UX differences that will be implemented separately (ref p91TBi-3f7-p2) - navigation CTA, ellipsis menu, product type row etc.

![simulator](https://user-images.githubusercontent.com/1945542/92909428-ad996780-f459-11ea-95aa-9c49f9577f7e.gif)

### Product type bottom sheet

light | dark
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-09-11 at 17 58 35](https://user-images.githubusercontent.com/1945542/92908484-d79e5a00-f458-11ea-851c-34481fb4fece.png) | ![Simulator Screen Shot - iPhone 11 - 2020-09-11 at 17 59 03](https://user-images.githubusercontent.com/1945542/92908499-dbca7780-f458-11ea-8eee-8aff19a96500.png)


Update release notes:
